### PR TITLE
Update LIMS constants and api to comply with changes made to LIMS

### PR DESCRIPTION
**Pull request by Karl-Svard**
## Description
The LIMS constants file has been out of date for quite some time. Changes was made to certain steps in LIMS, in particular those that went from using AM to Atlas method docs, and this was never corrected in cg. The result was that our delivery reports showed incorrect information. 

### Fixed
- updates the lims constants file to be in sync with the current state of lims
- make the get method function in the lims api be compatible with both AM and Atlas method documentation

### How to prepare for test
- [ ] ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] paxa the environment: `paxa`
- [ ] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan
- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Comment by Karl-Svard** [comment]
Hi @ivadym, could you help me test these changes aswell?

**Review by ivadym — COMMENTED** [review]
Well done 💯 

I just have some minor comments :)

And yes, we can test it generating some delivery reports

File: cg/apps/lims/api.py [hunk 1]
```diff
@@ -344,21 +353,35 @@ def _get_methods(self, step_names_udfs, lims_id):
...
     @staticmethod
-    def get_method_number(artifact, udf_key_number):
+    def get_method_type(artifact, method_udfs):
+        """
+        Assess which type of method documentation has been used, AM or Atlas.
+        """
+        if "atlas_version" in method_udfs and artifact.parent_process.udf.get(method_udfs["atlas_version"]):
+            return "Atlas"
+        else:
+            return "AM"
```
[the remark this answers is not in the document]
Comment (Karl-Svard) [reply cg/apps/lims/api.py:377]: That sounds like a good idea! Will add it and make appropiate changes to places where I have used method doc type 👍

**Comment by ivadym** [comment]
### Tests on Stage:

Original delivery report:
Screen Shot 2022-10-07 at 14 20 24

Fix:
Screen Shot 2022-10-07 at 14 43 10

> TODO
> Testing a custom sequencing type in Atlas (should overwrite the AM one):

**Comment by ivadym** [comment]
### Tests on Stage:

<details>
 <summary>Original delivery report:</summary>
Screen Shot 2022-10-07 at 14 20 24
</details>

<details>
 <summary>Fix:</summary>
Screen Shot 2022-10-07 at 14 43 10
</details>

<details>
 <summary>Testing a custom sequencing type in Atlas (should overwrite the AM one):</summary>
Screen Shot 2022-10-14 at 13 41 13
</details>

File: cg/apps/lims/api.py [hunk 2]
```diff
@@ -322,20 +323,32 @@ def _most_recent_date(self, dates: list):
...
 
         for process_name in step_names_udfs:
             artifacts = self.get_artifacts(process_type=process_name, samplelimsid=lims_id)
-            if artifacts:
-                udf_key_number = step_names_udfs[process_name]["method_number"]
-                udf_key_version = step_names_udfs[process_name]["method_version"]
+            if not artifacts:
+                continue
+            # Get a list of parent processes for the artifacts
+            processes = self.get_processes_from_artifacts(artifacts=artifacts)
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:336]: type hint

File: cg/apps/lims/api.py [hunk 3]
```diff
@@ -322,20 +323,32 @@ def _most_recent_date(self, dates: list):
...
+            for process in processes:
+                # Check which type of method document has been used
+                method_type = self.get_method_type(process, step_names_udfs[process_name])
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:339]: Type hint

**Review by ivadym — APPROVED** [review]
Great work 💯 🔥

File: cg/apps/lims/api.py [hunk 4]
```diff
@@ -322,20 +323,32 @@ def _most_recent_date(self, dates: list):
...
+                if method_type == DocumentationMethod.ATLAS:
+                    udf_key_method_doc = step_names_udfs[process_name]["atlas_document"]
+                    udf_key_version = step_names_udfs[process_name]["atlas_version"]
+                else:
+                    udf_key_method_doc = step_names_udfs[process_name]["method_number"]
+                    udf_key_version = step_names_udfs[process_name]["method_version"]
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:345]: I think this look up should be a function as is it traverses 2 levels of of a container i.e. it requires knowledge on what the internals of step_names_udfs looks like

File: cg/apps/lims/api.py [hunk 5]
```diff
@@ -344,28 +357,58 @@ def _get_methods(self, step_names_udfs, lims_id):
...
 
         return None
 
     @staticmethod
-    def get_method_number(artifact, udf_key_number):
+    def get_processes_from_artifacts(artifacts: List[Artifact]) -> List[Process]:
+        """
+        Get a list of parent processes from a set of given artifacts.
+        """
+        processes = []
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:382]: Type hint

File: cg/apps/lims/api.py [hunk 6]
```diff
@@ -344,28 +357,58 @@ def _get_methods(self, step_names_udfs, lims_id):
...
+        for artifact in artifacts:
+            parent_process = artifact.parent_process
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:384]: Type hint

File: cg/apps/lims/api.py [hunk 7]
```diff
@@ -344,28 +357,58 @@ def _get_methods(self, step_names_udfs, lims_id):
...
+        return processes
+
+    @staticmethod
+    def get_method_type(process: Process, method_udfs: Dict) -> str:
+        """
+        Assess which type of method documentation has been used, AM or Atlas.
+        """
+        if "atlas_version" in method_udfs and process.udf.get(method_udfs["atlas_version"]):
+            return DocumentationMethod.ATLAS
+
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:396]:
```suggestion
```

File: cg/apps/lims/api.py [hunk 8]
```diff
@@ -344,28 +357,58 @@ def _get_methods(self, step_names_udfs, lims_id):
...
+        return DocumentationMethod.AM
+
+    @staticmethod
+    def get_method_document(process: Process, udf_key_method_doc: str) -> str:
         """
         get method number for artifact
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:402]:
```suggestion
        Return method number for artifact.
```

File: cg/apps/lims/api.py [hunk 9]
```diff
@@ -344,28 +357,58 @@ def _get_methods(self, step_names_udfs, lims_id):
...
         """
-        return artifact.parent_process.udf.get(udf_key_number)
+        return process.udf.get(udf_key_method_doc)
 
     @staticmethod
-    def get_method_version(artifact, udf_key_version):
+    def get_method_version(process: Process, udf_key_version: str) -> str:
         """
         get method version for artifact
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:409]:
```suggestion
        Return method version for artifact.
```

**Review by henrikstranneheim — APPROVED** [review]
Well done!

File: cg/apps/lims/api.py (hunk 2 above)
Comment (Karl-Svard) [reply cg/apps/lims/api.py:336]: 👍

File: cg/apps/lims/api.py (hunk 3 above)
Comment (Karl-Svard) [reply cg/apps/lims/api.py:339]: 👍

File: cg/apps/lims/api.py (hunk 5 above)
Comment (Karl-Svard) [reply cg/apps/lims/api.py:382]: 👍

File: cg/apps/lims/api.py (hunk 7 above)
Comment (Karl-Svard) [reply cg/apps/lims/api.py:396]: 👍

File: cg/apps/lims/api.py (hunk 8 above)
Comment (Karl-Svard) [reply cg/apps/lims/api.py:402]: 👍

File: cg/apps/lims/api.py (hunk 9 above)
Comment (Karl-Svard) [reply cg/apps/lims/api.py:409]: 👍

File: cg/apps/lims/api.py (hunk 4 above)
Comment (Karl-Svard) [reply cg/apps/lims/api.py]: I added a new function for these lines, get_method_udf_values. Was it this you meant?

**Comment by henrikstranneheim** [comment]
@Karl-Svard Just 2 code smells to fix

**Comment by Karl-Svard** [comment]
### Deployed
```Shell
Successfully built cg
Installing collected packages: cg
  Attempting uninstall: cg
    Found existing installation: cg 22.85.2
    Uninstalling cg-22.85.2:
      Successfully uninstalled cg-22.85.2
Successfully installed cg-22.85.3
WARNING: You are using pip version 21.3.1; however, version 22.3 is available.
You should consider upgrading via the '/home/proj/bin/conda/envs/P_main/bin/python -m pip install --upgrade pip' command.
remote: Enumerating objects: 1677, done.
remote: Counting objects: 100% (639/639), done.
remote: Compressing objects: 100% (23/23), done.
remote: Total 1677 (delta 629), reused 618 (delta 616), pack-reused 1038
Receiving objects: 100% (1677/1677), 238.75 KiB | 0 bytes/s, done.
Resolving deltas: 100% (1329/1329), completed with 98 local objects.
From https://github.com/Clinical-Genomics/cg
   63450a3..ddc3ecf  master     -> origin/master
   168292c..52a60c7  add-balsamic-loqusdb-cli -> origin/add-balsamic-loqusdb-cli
 * [new branch]      fix-missing-genepanels-in-scout -> origin/fix-missing-genepanels-in-scout
   2c7b336..7b06b98  rnafusion  -> origin/rnafusion
   2546df0..14e6ed3  split-observations-api -> origin/split-observations-api
 * [new tag]         v22.85.3   -> v22.85.3
Updating 63450a3..ddc3ecf
Fast-forward
 .bumpversion.cfg     |  2 +-
 cg/__init__.py       |  2 +-
 cg/apps/lims/api.py  | 92 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++--------------------
 cg/constants/lims.py | 32 +++++++++++++++++++++++++++++++-
 setup.py             |  2 +-
 5 files changed, 106 insertions(+), 24 deletions(-)
/home/hiseq.clinical
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/hiseq.clinical
done.
Log deploy... done.
Running --version on each Clinical-Genomics developed tool to verify that all requirements are met:
git-lint 0.1.2 requires termcolor, which is not installed.
balsamic 6.0.0 requires colorclass, which is not installed.
argcomplete 1.12.1 has requirement importlib-metadata<3,>=0.23, but you have importlib-metadata 4.11.3.
cg, version 22.85.3
housekeeper, version 3.2.0
demux, version 5.10.4
cglims, version 1.3.4
2d [hiseq.clinical@hasta:~] [P_main] 2m26s $ cg --version
cg, version 22.85.3
```

**Pull request merged**